core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthServiceBuilder.java

```diff
@@ -0,0 +1,109 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import com.google.common.collect.Lists;
+
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.server.Service;
+
+/**
+ * Builds a new {@link HttpAuthService}.
+ */
+public final class HttpAuthServiceBuilder {
+
+    private final List<Predicate<? super HttpHeaders>> predicates = new ArrayList<>();
+
+    /**
+     * Adds an authorization predicate.
+     */
+    public HttpAuthServiceBuilder add(Predicate<? super HttpHeaders> predicate) {
+        predicates.add(predicate);
+        return this;
+    }
+
+    /**
+     * Adds multiple authorization predicates.
+     */
+    public HttpAuthServiceBuilder addAll(Iterable<? extends Predicate<? super HttpHeaders>> predicates) {
+        this.predicates.addAll(Lists.newArrayList(predicates));
+        return this;
+    }
+
+    /**
+     * Adds an HTTP basic authorization predicate.
+     */
+    public HttpAuthServiceBuilder addBasicAuth(Predicate<? super BasicToken> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final BasicTokenExtractor extractor = new BasicTokenExtractor();
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                BasicToken token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Adds an OAuth1a authorization predicate.
+     */
+    public HttpAuthServiceBuilder addOAuth1a(Predicate<? super OAuth1aToken> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final OAuth1aTokenExtractor extractor = new OAuth1aTokenExtractor();
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                OAuth1aToken token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Adds an OAuth2 authorization predicate.
+     */
+    public HttpAuthServiceBuilder addOAuth2(Predicate<? super OAuth2Token> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final OAuth2TokenExtractor extractor = new OAuth2TokenExtractor();
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                OAuth2Token token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpAuthService} {@link Service} decorator that supports all of given
+     * authorization {@link Predicate}s.
+     */
+    public Function<Service<? super HttpRequest, ? extends HttpResponse>, HttpAuthService> build() {
```

Comment:
You mean, renaming `build()` into `newDecorator()`? If it is the case, it would be better to rename `HttpAuthServiceBuilder` into `HttpAuthDecoratorBuilder`.